core/clipboard.ts

```diff
@@ -68,23 +108,35 @@ export function paste(): ICopyable | null {
  *     duplication failed.
  * @internal
  */
-export function duplicate(toDuplicate: ICopyable): ICopyable | null {
+export function duplicate<
+  U extends ICopyData,
+  T extends ICopyable<U> & IHasWorkspace,
+>(toDuplicate: T): T | null {
   return TEST_ONLY.duplicateInternal(toDuplicate);
 }
 
 /**
  * Private version of duplicate for stubbing in tests.
  */
-function duplicateInternal(toDuplicate: ICopyable): ICopyable | null {
-  const oldCopyData = copyData;
-  copy(toDuplicate);
-  if (!copyData || !source) return null;
-  const pastedThing =
-    globalRegistry
-      .getObject(globalRegistry.Type.PASTER, copyData.paster, false)
-      ?.paste(copyData, source) ?? null;
-  copyData = oldCopyData;
-  return pastedThing;
+function duplicateInternal<
+  U extends ICopyData,
+  T extends ICopyable<U> & IHasWorkspace,
+>(toDuplicate: T): T | null {
+  const oldCopyData = stashedCopyData;
+  const oldWorkspace = stashedWorkspace;
+
+  const data = copy(toDuplicate);
+
+  // I hate side effects.
+  stashedCopyData = oldCopyData;
```

Comment:
Could you just call `toDuplicate.toCopyData()` instead of calling `copy()`? Then you wouldn't have to save and reset the stashed data.

Reply:
Yeshhh thank you!